hibernate-core/src/main/java/org/hibernate/engine/internal/EntityEntryContext.java

```diff
@@ -396,10 +419,81 @@ public static EntityEntryContext deserialize(ObjectInputStream ois, StatefulPers
 		return context;
 	}
 
+	private static EntityEntry deserializeEntityEntry(char[] entityEntryClassNameArr, ObjectInputStream ois, StatefulPersistenceContext rtn){
+		EntityEntry entry = null;
+
+		final String entityEntryClassName = new String( entityEntryClassNameArr );
+		final Class entityEntryClass =   rtn.getSession().getFactory().getServiceRegistry().getService( ClassLoaderService.class ).classForName( entityEntryClassName );
+
+		try {
+			final Method deserializeMethod = entityEntryClass.getDeclaredMethod( "deserialize", ObjectInputStream.class,	PersistenceContext.class );
+			entry = (EntityEntry) deserializeMethod.invoke( null, ois, rtn );
+		}
+		catch (NoSuchMethodException e) {
+			log.errorf( "Enable to deserialize [%s]", entityEntryClassName );
+		}
+		catch (InvocationTargetException e) {
+			log.errorf( "Enable to deserialize [%s]", entityEntryClassName );
+		}
+		catch (IllegalAccessException e) {
+			log.errorf( "Enable to deserialize [%s]", entityEntryClassName );
+		}
+
+		return entry;
+
+	}
+
 	public int getNumberOfManagedEntities() {
 		return count;
 	}
 
+	/*
+	Check instance type of EntityEntry and if type is ImmutableEntityEntry, check to see if entity is referenced cached in the second level cache
+	 */
+	private boolean canClearEntityEntryReference(ManagedEntity managedEntity){
+
+		if(managedEntity.$$_hibernate_getEntityEntry() == null) {
+			return true;
+		}
+
+		if(!(managedEntity.$$_hibernate_getEntityEntry() instanceof ImmutableEntityEntry)) {
+			return true;
+		}
+		else {
+			//need to check for reference caching in second level cache here
+			CacheEntry entry = null;
```

Comment:
Oh wow.  No :)

You physically look up the thing from the cache in order to tell if it is a cache-by-reference?  That's a very expensive way to do this.  The EntityPersister knows this: org.hibernate.persister.entity.EntityPersister#canUseReferenceCacheEntries  And actually, since this would be getting used from such critical path code, I'd even suggest we optimize the internals of EntityPersister#canUseReferenceCacheEntries